core/arch/opnd.h

```diff
@@ -2350,19 +2433,36 @@ DR_API
 /**
  * Assumes that \p reg is a DR_REG_ constant.
  * Returns true iff it refers to an xmm (128-bit SSE/SSE2) register
- * or a ymm (256-bit multimedia) register.
+ * or a ymm (256-bit multimedia) or a zmm (512-bit multimedia) register.
  */
 bool
 reg_is_xmm(reg_id_t reg);
 
 DR_API
 /**
  * Assumes that \p reg is a DR_REG_ constant.
- * Returns true iff it refers to a ymm (256-bit multimedia) register.
+ * Returns true iff it refers to a ymm (256-bit multimedia) or a zmm
+ * (512-bit multimedia) register.
```

Comment:
Don't add zmm here: keep it what you'd expect, just ymm.

Reply:
Why.
reg_is_xmm => xmm, ymm, zmm. 
reg_is_ymm => ymm. zmm.
reg_is_strictly_xmm => xmm.
reg_is_strictly_ymm => ymm.

Otherwise it'll be total confusion. The reg_is_* functions currently include their wider counter parts.

Reply:
Should we add a "is_strictly_zmm" as well? Just to be complete?

Reply:
Actually, let's just _only_ add reg_is_strictly_zmm.

Reply:
I meant to imply that I thought an "is_strictly_zmm" version would exist, in addition to strictly xmm and strictly ymm. I am OK with just adding "is_strictly_zmm" in this CL, leaving the others as-is for this PR, and seeing what we need as we move forward.

I also wonder about the usecases: what check will users most often want?

Reply:
> I also wonder about the usecases: what check will users most often want?

Right, I would look at the existing reg_is_xmm case: if there are a lot, maybe what we need is a reg_is_simd() that does not include MMX registers: not sure what to call it.  Basically what reg_is_xmm did, but a better name.  reg_is_nonfp_simd(), reg_is_simd128plus()?  Both bad.  W/o a new routine we force everyone to call "reg_is_simd(r) && !reg_is_mmx(r)".

Reply:
Are we good here now? reg_is_xmm, reg_is_ymm has been marked with \deprecated, reg_is_strictly_[xmm|ymm|zmm] was added.

Reply:
The question from the prior 2 comments above does not seem to be resolved: Deciding what will existing callers of reg_is_xmm be switched to?

Reply:
If that should be part of this PR - my judgment would be it's not - my suggestion then is

- deprecate reg_is_xmm (as in this PR already - because it includes ymm)
- possibly deprecate reg_is_simd? (I don't see many common uses to query MMX or SEE+AVX)?
- add reg_is_strictly_xmm(), reg_is_strictly_ymm(), reg_is_strictly_zmm()
- add reg_is_avx() (returns xmm || ymm || zmm)
- optionally add reg_is_sse() (returns reg_is_xmm())

Existing callers then will either call reg_is_strictly_xmm(), reg_is_avx().

Opinions?

Reply:
Not nec. part of this PR, but it's a bad idea to make a design decision on the deprecate w/o a plan in place for how to update existing uses: we might decide we have to change the deprecation based on the update options.

Reply:
Not a fan of having the names coming from the ISA extensions: I can never remember which extension added what, but I can certainly remember xmm vs mmx.  I would prefer names based on the registers and not on the names of the extensions where they were added.

Reply:
Plus, it may be confusing that XMM0, which was not add by AVX*, would return true for reg_is_avx().

Reply:
For me the pity is really that the names "reg_is_xmm|ymm" are already taken. Which is why I originally suggested to keep their meaning and even add "|| zmm" to reg_is_ymm(). And then add "reg_is_strictly_xmm|ymm|zmm". That was my plan A. Most likely, existing uses of reg_is_xmm would have been left as-is. In this review it was established to not do that, which is ok by the way.  But now We want to deprecate "reg_is_xmm" which is why I suggested above, which is mainly to add reg_is_avx(), or alternative name, reg_is_sse_avx().

Opinions?

Reply:
Are we settling on this one? I made 2 suggestions

Reply:
What did the requested existing use case survey say?  A quick grep shows a bunch of these: `reg_is_xmm(reg) || reg_is_mmx(reg)`, which would become `reg_is_simd`.  In Dr. Memory I see `reg_is_xmm(reg) && !reg_is_ymm(reg)` among others.

Reply:
I assume you did a more thorough examination.  But maybe there is no need for anything beyond reg_is_simd and the series of reg_is_{mmx,xmm,ymm,zmm} (whatever their names are: the ones that only identify that class).

Reply:
reg_is_xmm(reg) || reg_is_mmx(reg) would become reg_is_simd()
reg_is_xmm(reg) && !reg_is_ymm(reg) would become reg_is_strictly_xmm()
reg_is_xmm(reg) would eventually become reg_is_avx() or if not desired, then reg_is_strictly_xmm() || reg_is_strictly_zmm() || reg_is_strictly_zmm() 

My suggestion was to add reg_is_avx() or reg_is_sse() or reg_is_sse_avx(). Will this work with Dr. Memory?

Reply:
As mentioned I was not a big fan of using the ISA extension name ("avx" or "sse"), nor of having XMM0 return true for reg_is_avx().  I don't have a great alternative proposed name though.  Maybe nothing should be added (or else something cross-platform along the lines of reg_is_simd()).